src/common/cuda_utils.cc

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2019 by Contributors
+ * \file cuda_utils.cc
+ * \brief CUDA debugging utilities.
+ */
+
+#include <mxnet/base.h>
+#include "cuda_utils.h"
+
+#if MXNET_USE_CUDA == 1
+
+namespace mxnet {
+namespace common {
+namespace cuda {
+
+// The oldest version of cuda used in upstream MXNet CI testing, both for unix and windows.
+// Users that have rebuilt MXNet against older versions will we advised with a warning to upgrade
+// their systems to match the CI level.  Minimally, users should rerun the CI locally.
+#if defined(_MSC_VER)
+#define MXNET_CI_OLDEST_CUDA_VERSION  9020
+#else
+#define MXNET_CI_OLDEST_CUDA_VERSION 10000
+#endif
+
+// Dynamic init here will emit a warning if runtime and compile-time cuda lib versions mismatch.
+// Also if the user has recompiled their source to a version no longer tested by upstream CI.
+bool cuda_version_check_performed = []() {
+  // Don't bother with checks if there are no GPUs visible (e.g. with CUDA_VISIBLE_DEVICES="")
+  if (dmlc::GetEnv("MXNET_CUDA_VERSION_CHECKING", true) && Context::GetGPUCount() > 0) {
+    int linkedAgainstCudaVersion = 0;
+    CUDA_CALL(cudaRuntimeGetVersion(&linkedAgainstCudaVersion));
+    if (linkedAgainstCudaVersion != CUDA_VERSION)
+      LOG(WARNING) << "cuda library mismatch: linked-against version " << linkedAgainstCudaVersion
```

Comment:
Just want to make sure I'm understanding this one.  If a user runs with CUDA 10.2, but the library was linked against 10.1 would this issue a warning?  I tend to do that fairly often, is it against best practices?

Reply:
So 'yes' there would be a warning if the user built against 10.1, but ran with 10.2.  These warnings can be turned off with an environment variable setting MXNET_CUDA_VERSION_CHECKING=0.  The idea behind the 'advisory' is that the user may want to rebuild to get the new functionality present in 10.2, or perhaps to avoid work-arounds for any issues of 10.1.  It's probably more useful with the CUDNN version checks, where we have far more compile guards based on version minor numbers.  Do you feel these warnings would be unwelcome to users?

Reply:
I think the question is what can be the issues when linking against a smaller cuda, leaving performance gains on the table? I think you guys are the experts, I was getting some info from here: https://docs.nvidia.com/deploy/cuda-compatibility/#binary-compatibility 
Does this warning indicate a real problem or will it confuse users, when there's nothing wrong on running with a newer cuda.

Reply:
After studying the link supplied by @larroy (thanks!), I need to retract what I said above.  Based on a new understanding, I have removed the runtime check of the cuda-runtime library version.  The check is unnecessary since (per the link) the major.minor of the cuda runtime must match for the libmxnet.so lib to load.  It was instructive for me to do a `ldd libmxnet.so`:
```
	libcudart.so.9.2 => /usr/local/cuda/lib64/libcudart.so.9.2 (0x00007f361cc64000)
	libcudnn.so.7 => /usr/lib/x86_64-linux-gnu/libcudnn.so.7 (0x00007f35f9cdf000)
	libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007f35f3703000)
```
Note the extra '.minor' number on libcudart.so.  So while a compiled-against cudnn 7.2, might run against a cudnn 7.6, a compiled-against cuda 10.1 won't run against a cuda 10.2.  Now, keep in mind we're talking about the cuda `runtime` library, so libcudart.so as set up by the toolkit install.  Your experience on 10.1 vs. 10.2 @KellenSunderland was probably based on upgrading the driver to a higher version, while leaving the toolkit install the same.

Let me know if the PR is now to your liking.  I've left in the test of the cuda runtime version against the threshold MXNET_CI_OLDEST_CUDA_VERSION.  The idea is that once we no longer test against a particular cuda version, then bugs will creep in with new PRs.  We'd prefer users to not be on the front line of bug finding, so we should encourage them to upgrade.

Back to your original question- there will be no warning for upgrading the driver to a newer version (e.g. 10.2) while leaving the toolkit at 10.1

Reply:
Thanks for the clarification, I think your change brings value to align users and what's tested in CI/CD.

Reply:
Thanks for the updates and clarification Dick.